### PR TITLE
Fully deploy a cobbler node

### PR DESCRIPTION
**Pull request by zmc**
I believe this does everything we need to deploy a new cobbler node.

Well, one thing it does *not* do is deploy ``/root/.vault_pass.txt`` - because it can't really ;)

Feedback?

### CODE REVIEW

File: roles/cobbler/defaults/main.yml [hunk 1]
```diff
@@ -4,6 +4,21 @@ iso_dir: /var/lib/cobbler/isos
 # Mount point to use for ISOs during import
 iso_mount: /mnt/iso
 
+etc_files:
+ - users.digest
+
+settings:
```
Comment (andrewschoen) [inline roles/cobbler/defaults/main.yml]: The actual setting could be the key here instead of 'name' and 'value'.  Then you could do something like ``settings.signature_url`` in usage.

File: roles/cobbler/tasks/apt_systems.yml [hunk 2]
```diff
@@ -0,0 +1,31 @@
+---
+- name: Install cobbler
+  apt:
+    name: "{{ cobbler_package }}"
+    state: latest
+  register: install_cobbler
+
+- name: Install syslinux
```
Comment (andrewschoen) [inline roles/cobbler/tasks/apt_systems.yml]: Should we put these extra packages in a list and use with_items to install them?  These next few tasks seem very repetitive.

File: roles/cobbler/tasks/fetch_cm_repos.yml [hunk 3]
```diff
@@ -8,3 +8,9 @@
   git: repo={{ secrets_repo.url }}
        dest=/root/{{ secrets_repo.name }}
        accept_hostkey=true
+
+- name: Symlink /etc/ansible/hosts
+  file: src=/root/{{ secrets_repo.name }}/ansible/inventory/ dest=/etc/ansible/hosts state=link force=yes
```
Comment (andrewschoen) [inline roles/cobbler/tasks/fetch_cm_repos.yml]: nitpick: can we put src, dest, state and force each on their own lines?

I find that format much easier to read and it's makes more friendly diffs when changing a task that's formatted that way.

File: roles/cobbler/tasks/fetch_cm_repos.yml [hunk 4]
```diff
@@ -8,3 +8,9 @@
...
+
+- name: Symlink /etc/ansible/secrets
+  file: src=/root/{{ secrets_repo.name }}/ansible/secrets/ dest=/etc/ansible/secrets state=link force=yes
```
Comment (andrewschoen) [inline roles/cobbler/tasks/fetch_cm_repos.yml]: nitpick: same formatting concern here.

File: roles/cobbler/tasks/main.yml [hunk 5]
```diff
@@ -1,4 +1,69 @@
 ---
+- name: Include cobbler keys.
+  include_vars: "{{ secrets_path | mandatory }}/cobbler_keys.yml"
```
Comment (andrewschoen) [inline roles/cobbler/tasks/main.yml]: I haven't seen cobbler_keys.yml yet, but I'm wondering if whatever is there could just be in the group_vars/cobbler.yml in the *-secrets repos.

Is this file encrypted?  I suppose that wouldn't work then.

File: roles/cobbler/tasks/main.yml [hunk 6]
```diff
@@ -1,4 +1,69 @@
...
+  no_log: true
+  tags:
+    - vars
+
+- name: Create /root/.ssh
+  file: path=/root/.ssh mode=700 state=directory
```
Comment (andrewschoen) [inline roles/cobbler/tasks/main.yml]: formatting nitpick

File: roles/cobbler/tasks/main.yml [hunk 7]
```diff
@@ -1,4 +1,69 @@
...
+  when: ansible_pkg_mgr == "apt"
+
+- name: Start cobbler
+  service:
+    name: "{{ cobbler_service }}"
+    state: started
+    enabled: yes
+
+- name: Enable tftpd
+  lineinfile: dest=/etc/xinetd.d/tftp regexp=disable line="        disable                 = no"
```
Comment (andrewschoen) [inline roles/cobbler/tasks/main.yml]: formatting nitpick

File: roles/cobbler/tasks/settings.yml [hunk 8]
```diff
@@ -0,0 +1,17 @@
...
+- name: Set server value
+  # Escape the colon below so the line will parse
+  lineinfile: dest=/etc/cobbler/settings regexp="^server{{':'}}" line="server{{'\:'}} {% for setting in settings %}{% if setting.name == 'server' %}{{ setting.value }}{% endif %}{% endfor %}"
+  register: server_value
+
+- include: restart.yml
+  when: dynamic_settings|changed or server_value|changed
+
+- name: Update settings
+  command: cobbler setting edit --name={{ item.name }} --value={{ item.value}}
```
Comment (andrewschoen) [inline roles/cobbler/tasks/settings.yml]: ah, now I see why having 'name' and 'value' on settings is nice.

File: roles/cobbler/tasks/vars.yml [hunk 9]
```diff
@@ -0,0 +1,3 @@
+---
+- name: Include package type specific vars.
```
Comment (andrewschoen) [inline roles/cobbler/tasks/vars.yml]: Is this file needed?  I think you do this already in main.yml

File: roles/cobbler/tasks/yum_systems.yml [hunk 10]
```diff
@@ -0,0 +1,31 @@
+---
+- name: Install cobbler
+  yum:
+    name: "{{ cobbler_package }}"
+    state: latest
+  register: install_cobbler
+
+- name: Install syslinux
```
Comment (andrewschoen) [inline roles/cobbler/tasks/yum_systems.yml]: I think these install tasks should also be in a list of packages to install and then used with 'with_items'.

File: roles/cobbler/templates/kickstarts/cephlab_ubuntu.preseed [hunk 11]
```diff
@@ -89,7 +89,7 @@ d-i user-setup/encrypt-home boolean false
 #if $os_version == 'precise'
 d-i pkgsel/include string wget ntpdate bash sudo openssh-server
 #else
-d-i pkgsel/include string u-boot-tools pastebinit initramfs-tools wget linux-firmware linux-firmware-nonfree ntpdate bash devmem2 fbset sudo openssh-server udev-discover
+d-i pkgsel/include string u-boot-tools pastebinit initramfs-tools wget linux-firmware linux-firmware-nonfree ntpdate bash devmem2 fbset sudo openssh-server udev-discover gawk
```
Comment (andrewschoen) [inline roles/cobbler/templates/kickstarts/cephlab_ubuntu.preseed]: Does precise need gawk as well?

File: roles/cobbler_systems/tasks/populate_systems.yml [hunk 12]
```diff
@@ -0,0 +1,27 @@
...
+  command: cobbler system add --name={{ item.split('.')[0] }} --profile={{ default_profile }} --mac={{ hostvars[item].mac }} --ip-address={{ hostvars[item].ip }} --interface={{ hostvars[item].interface }} --kopts="{{ hostvars[item].kernel_options }}" --netboot-enabled false
+  with_items: cobbler_systems_add
+
+- name: set cobbler_systems_update
+  set_fact:
+    cobbler_systems_update:
+      "{{ groups.cobbler_managed | intersect(cobbler_systems_current) }}"
+
+- name: Update existing systems in cobbler
+  command: cobbler system edit --name={{ item.split('.')[0] }} --mac={{ hostvars[item].mac }} --ip-address={{ hostvars[item].ip }} --interface={{ hostvars[item].interface }} --kopts="{{ hostvars[item].kernel_options }}"
```
Comment (andrewschoen) [inline roles/cobbler_systems/tasks/populate_systems.yml]: Are you adding things like 'kernel_options' to the *-secrets repo for each host?  I assume things like mac, ip and interface are already there.

File: roles/cobbler/defaults/main.yml (hunk 1 above)
Comment (zmc) [inline roles/cobbler/defaults/main.yml]: eh, then I wouldn't be able to conveniently loop over them!

File: roles/cobbler/tasks/main.yml (hunk 5 above)
Comment (zmc) [inline roles/cobbler/tasks/main.yml]: It is; it contains private keys

File: roles/cobbler/templates/kickstarts/cephlab_ubuntu.preseed (hunk 11 above)
Comment (zmc) [inline roles/cobbler/templates/kickstarts/cephlab_ubuntu.preseed]: Oh! Probably!

File: roles/cobbler_systems/tasks/populate_systems.yml (hunk 12 above)
Comment (zmc) [inline roles/cobbler_systems/tasks/populate_systems.yml]: Yes, all that is correct

File: roles/cobbler/tasks/main.yml (hunk 5 above)
Comment (andrewschoen) [inline roles/cobbler/tasks/main.yml]: Ok, cool.  This is fine then.

File: roles/cobbler_systems/tasks/populate_systems.yml (hunk 12 above)
Comment (andrewschoen) [inline roles/cobbler_systems/tasks/populate_systems.yml]: I was suggesting it was wrong, I was just surprised to see that 'kernel_options' was something provided by default from ansible.

Comment (zmc) [inline roles/cobbler_systems/tasks/populate_systems.yml]: I'm confused by this comment :-/

Comment (andrewschoen) [inline roles/cobbler_systems/tasks/populate_systems.yml]: You're confused because of my terrible typos. I meant 'wasn't', not 'was', in that first sentence.

Having not seen the *-secrets PRs for this yet 'kernel_options' was confusing me, but that's all cleared up now.

**Comment by andrewschoen** [comment]
After having seen the associated PRs for the *-secrets repos I think you should make sure that all variables being used and defined in a *-secrets repo are listed and defaulted in defaults/main.yml as well.

I see these as added, but not listed there: interface, kernel_options, default_profile and ipmi_domain.

**Comment by zmc** [comment]
@andrewschoen Those vars are defined in ``roles/cobbler_systems/defaults/main.yml``, which I had conveniently forgotten to commit earlier. It's there now. Missing is ``ipmi_domain`` which is not currently used, but I'll fix that. IMO we should get this in and worry about ``ipmi_domain`` in the near future

File: roles/cobbler/tasks/main.yml [hunk 13]
```diff
@@ -1,4 +1,78 @@
...
+  when: ansible_pkg_mgr == "apt"
+
+- name: Start cobbler
+  service:
+    name: "{{ cobbler_service }}"
+    state: started
+    enabled: yes
+
+- name: Enable tftpd
+  lineinfile:
```
Comment (andrewschoen) [inline roles/cobbler/tasks/main.yml]: I think you will need 'backrefs: yes' here so that it won't add this line multiple times on subsequent runs

File: roles/cobbler/tasks/settings.yml [hunk 14]
```diff
@@ -0,0 +1,17 @@
+---
+- name: Enable dynamic settings modification
+  # Escape the colon below so the line will parse
+  lineinfile: dest=/etc/cobbler/settings regexp=^allow_dynamic_settings line="allow_dynamic_settings{{':'}} 1"
```
Comment (andrewschoen) [inline roles/cobbler/tasks/settings.yml]: formatting nitpick, also backrefs

File: roles/cobbler/tasks/settings.yml [hunk 15]
```diff
@@ -0,0 +1,17 @@
...
+  register: dynamic_settings
+
+- name: Set server value
+  # Escape the colon below so the line will parse
+  lineinfile: dest=/etc/cobbler/settings regexp="^server{{':'}}" line="server{{'\:'}} {% for setting in settings %}{% if setting.name == 'server' %}{{ setting.value }}{% endif %}{% endfor %}"
```
Comment (andrewschoen) [inline roles/cobbler/tasks/settings.yml]: formatting nickpick, also backrefs

**Comment by andrewschoen** [comment]
This is looking great!  Just a few small things and then I think we're good to merge.

File: roles/cobbler/tasks/main.yml (hunk 13 above)
Comment (zmc) [inline roles/cobbler/tasks/main.yml]: Maybe I don't fully understand backrefs, but it seems it isn't needed:
```
TASK: [cobbler | Enable tftpd] ************************************************
ok: [plana32.front.sepia.ceph.com] => {"backup": "", "changed": false, "msg": ""}

TASK: [cobbler | Reload xinetd] ***********************************************
skipping: [plana32.front.sepia.ceph.com]
```

Comment (andrewschoen) [inline roles/cobbler/tasks/main.yml]: If it's not an issue then that's awesome!  I was just remembering I ran into an issue with the testnodes role where if I was using 'regexp' and 'line' that 'line' would be added on every run if 'regexp' was found, even if 'line' already existed.  Adding 'backrefs' fixed that.
